### DISCUSSION TITLE
Get directory contents with lua

### DISCUSSION THREAD

**Discussion by callimero** [RomanKubiak/ctrlr#127] [Ideas]
Hi,

not sure if this is the right place for that question. I bought a Yamaha Reface CS and use a Reface CS panel for that. It works like a charm but I miss a thing(see below) and as soon I smell some editor and scripting language I start to fiddle around. 

I can load "patches" (sysex) files from disk using the Filewindow. Great! But for preview hundreds of sounds it would be very boring to load every single patch. So I thought

 "why not having a function to have a +/- buttons to switch between all patches in a disk directory?"

I digged around a bit but this is all a bit new to me. Is there a lua or support function to get a list of all files in a directory? I guess that would make it possible for me to add me these function.

Best,
Carsten

**Comment by callimero** [comment]
Dang! Just stumbled across the mention of "file list" in the source and found something...

t = {}
f = File ("c:\\devel")
f:findChildFiles (t, File.findDirectories, true, "*")
for i,v in ipairs(t) do
	console (v:getFullPathName())
end

**Comment by callimero** [comment]
Nothing works.... Well... at least not predictable

After having a big problem with 6.0.4 (CTRLR Modulators not reacting for incoming CCs) I am back to 5.3.

However, above code does wacky things also at 5.3. 

`
function myTest()
    console (String("Hallo Test!"))
    patches = {}
--    local f = File ("C:/tmp")       -- Works
    local f = File ("C:/tmp/SySEx")       -- Not Working. YES the folder is there

        f:findChildFiles(patches, File.findDirectories, false, "*")
        for i,v in ipairs(patches) do
            console (String(i))
            console (v:getFullPathName())
        end
end
`
Using c:/tmp as path works, directory content in console. Using C:/tmp/SySEx does not list anything. Yes the directory exist. I tested with more cases, it is not the pathlength, not the depth of the path. I mean c:/tmp works but c:/SySEx not?!

I am lost! Hope it is nothing totally simple which I caused...

**Comment by dobo365** [comment]
Hi,
  Use 5.3.201 untill there is another stable version announced (or closed to stable, Roman is working on 5.5.x and there is good hope).
About the File issues, I encounter sometimes the same stuff.
Look in Juce API documentation for the different functions available: [https://docs.juce.com/master/classFile.html](https://docs.juce.com/master/classFile.html)
Then you need to experiment a bit with different syntax (single quotes, double quotes, slash, backslash...)
Use a test function to verify directly in your console like isDirectory() until you get a nice "true"
`console(tostring(fd:isDirectory()))`

I had also some issue a few days ago but don't remember what but anyway it was similar to yours.
Will try to remember and come back
/goodweather

**Reply by callimero** [reply]
Thanks for your answer.

documentation is a bit scattered and that makes it double hard, lerning lua (which seems strange to me :) and ctrlr and and and at the same time.

However: Since I use the Directory chooser for setting the path my problems are gone. Must be some string issue...

For the download... could be better communicated what is what as the download link just puts you in a folder with different versions...

**Reply by callimero** [reply]
![grafik](https://user-images.githubusercontent.com/7713008/104703782-09ca0c00-5718-11eb-893f-9ac63969b688.png)

Beside the ugly graphics (sorry Martin! :) I can now choose a patch directory, the list will be filled, I can choose there a patch, it will be send to the synth, I can use PREV/Next to go to prev/next patch! Great!

There are many bugs/quirks to iron I guess but for me it is now really helpfull.

BTW: I failed to have the PREV/NEXT buttons to send say +1/-1 to a script and use that to de/increase the actual patch value... So I use two scripts here... :( The value the button send was constantly increasing with each click, any pointers what I need to do?

**Reply by shooking** [reply]
hi @callimero
Lua is likely your friend here - but you already know that.

An IncDecButton generates a -|+
It has a modulator value. And Ctrlr can be asked to run Lua methods on modulator change


When you create your file list you _could_ write some Lua to say
  Get me that IncDecButton's id
  set min/max value

At this point your inc/dec knows range of values it can span.

So now when you interact with the IncDecButton it can set the Index of the comboBox.
And that _ought_ to trigger even "user selected a new index into the comboBox".
Which in turn ought to do the needful (load next patch).

Is that kind of what you want? I am sure there are better ways to do it but it's how I did it on my K1r panel.

Useful tips

In main Ctrlr panel in Edit mode there is a Utility button.
Select a control in your panel, then click the Utility.
Now you have the XML for both that modulator and its component.

Now you use set/get methods to alter this dynamically.
A comboBox has (taken from my Wave table drop down in K1r panel).
```
<?xml version="1.0" encoding="UTF-8"?>

<modulator modulatorVstExported="1" modulatorMax="255" vstIndex="376" modulatorIsStatic="0"
           modulatorGlobalVariable="-1.00000000000000000000" modulatorMuteOnStart="0"
           modulatorMute="0" modulatorExcludeFromSnapshot="0" modulatorValueExpression="modulatorValue+768"
           modulatorValueExpressionReverse="midiValue-768" modulatorControllerExpression="value"
           luaModulatorGetValueForMIDI="-- None" luaModulatorGetValueFromMIDI="-- None"
           modulatorLinkedToPanelProperty="-- None" modulatorLinkedToModulatorProperty="-- None"
           modulatorLinkedToModulator="-- None" modulatorLinkedToModulatorSource="1"
           modulatorLinkedToComponent="0" modulatorBaseValue="0" modulatorCustomIndex="0"
           modulatorCustomName="" modulatorCustomIndexGroup="0" modulatorCustomNameGroup=""
           modulatorVstNameFormat="%n" luaModulatorValueChange="-- None"
           name="WvName2" modulatorMin="0" modulatorValue="4">
  <midi midiMessageType="5" midiMessageChannelOverride="0" midiMessageChannel="1"
        midiMessageCtrlrNumber="1.00000000000000000000" midiMessageCtrlrValue="0"
        midiMessageMultiList="" midiMessageSysExFormula="F0 40 0Y 10 00 03 23 MS LS F7"/>
  <component componentLabelPosition="top" componentLabelJustification="centred"
             componentLabelHeight="14" componentLabelWidth="0" componentLabelVisible="1"
             componentLabelAlwaysOnTop="1" componentSentBack="0" componentLabelColour="0xff000000"
             componentLabelFont="&lt;Sans-Serif&gt;;12;0;0;0;0;1;3" componentVisibleName="WaveList2"
             componentMouseCursor="2" componentGroupName="WvGp2" componentGroupped="1"
             componentSnapSize="0" componentIsLocked="0" componentDisabled="0"
             componentRadioGroupId="0" componentRadioGroupNotifyMidi="1" componentVisibility="1"
             componentEffect="0" componentEffectRadius="1.00000000000000000000"
             componentEffectColour="0xff000000" componentEffectOffsetX="0"
             componentEffectOffsetY="0" componentExcludedFromLabelDisplay="0"
             componentValueDecimalPlaces="0" componentLuaMouseMoved="-- None"
             componentLuaMouseDown="-- None" componentLuaMouseDrag="-- None"
             componentLuaMouseDoubleClick="-- None" uiComboArrowColour="0xff0000ff"
             uiComboOutlineColour="0xff0000ff" uiComboTextColour="0xff000000"
             uiComboTextJustification="centred" uiComboFont="&lt;Sans-Serif&gt;;14;0;0;0;0;1;3"
             uiComboMenuFont="&lt;Sans-Serif&gt;;16;0;0;0;0;1;3" uiComboButtonColour="0xff0000ff"
             uiComboBgColour="ff56c312" uiComboMenuBackgroundColour="ff56c312"
             uiComboMenuFontColour="0xff000000" uiComboMenuHighlightColour="ffadd8e6"
             uiComboMenuFontHighlightedColour="0xff232323" uiComboContent="SIN 1st&#10;SIN 2nd&#10;SIN 3rd&#10;SIN 4th&#10;SIN 5th&#10;SIN 6th&#10;SIN 7th&#10;SIN 8th&#10;SIN 9th&#10;SIN 10th&#10;SIN 11th&#10;SIN 12th&#10;SIN 16th&#10;SAW 1&#10;SAW 2&#10;SAW 3&#10;SAW 4&#10;SAW 5&#10;SAW 6&#10;SAW 7&#10;SAW 8&#10;SAW 9&#10;SAW 10&#10;SAW 11&#10;SAW 12&#10;SAW 13&#10;SAW 14&#10;SAW 15&#10;SAW 16&#10;SAW 17&#10;SAW 18&#10;SAW 19&#10;SQUARE 1&#10;SQUARE 2&#10;SQUARE 3&#10;SQUARE 4&#10;SQUARE 5&#10;INVERSE-SAW&#10;TRIANGLE&#10;RANDOM&#10;FRENCH HORN&#10;STRING&#10;STRING&#10;STRING PAD&#10;PIANO 1&#10;EL. GRAND&#10;E. PIANO 1&#10;E. PIANO 2&#10;E. PIANO 3&#10;CLAVI&#10;VIBE&#10;A. GUITAR&#10;F. GUITAR&#10;F. GUITAR&#10;Ac BASS&#10;Ac BASS&#10;DIGI BASS 1&#10;PICK BASS&#10;DIGI BASS 2&#10;ROUND BASS&#10;FRETLESS&#10;FRETLESS&#10;FLUTE&#10;PANFLUTE&#10;HARMONICA&#10;GLOCKEN&#10;TINE&#10;HARP&#10;MARIMBA&#10;E. TOM&#10;LOG DRUM&#10;JASS ORGAN 1&#10;MELLO PAD&#10;SYNTH SOLO&#10;SYNTH 2&#10;FRENCH HORN&#10;FRENCH HORN&#10;BRASS&#10;BRASS&#10;BRASS&#10;BRASS&#10;TRUMPET&#10;TRUMPET&#10;VIOLIN&#10;STRING&#10;PIANO 1&#10;PIANO 2&#10;PIANO 3&#10;PIANO 2&#10;PIANO 3&#10;PIANO 4&#10;PIANO 4&#10;EL. GRAND&#10;E. PIANO 1&#10;E. PIANO 2&#10;E. PIANO 2&#10;CLAVI&#10;HARPSICHORD&#10;VIBE&#10;A. GUITAR&#10;F. GUITAR&#10;STRAT&#10;STRAT&#10;Ac BASS&#10;PULL BASS&#10;PULL BASS&#10;ROUND BASS&#10;SLAP BASS&#10;SLAP BASS&#10;SLAP BASS&#10;FRETLESS&#10;FRETLESS&#10;SYNTH BASS&#10;SYNTH BASS&#10;HARMONICA&#10;CLARINET&#10;CLARINET&#10;OBOE&#10;OBOE&#10;SHAKUHACHI&#10;ORIENTAL BELL&#10;ORIENTAL BELL&#10;BELL&#10;KOTO&#10;SITAR&#10;E. TOM&#10;LOG DRUM&#10;LOG DRUM&#10;STEEL DRUM&#10;STEEL DRUM&#10;VOICE 1&#10;VOICE 2&#10;ACCORDION&#10;ACCORDION&#10;JAZZ ORGAN 2&#10;ROCK ORGAN 1&#10;DRAW BAR 1&#10;DRAW BAR 2&#10;PIPE ORGAN 1&#10;PIPE ORGAN 2&#10;ROCK ORGAN 2&#10;SYNTH SOLO&#10;SYNTH SOLO&#10;SYNTH 2&#10;SYNTH 2&#10;SYNTH 3&#10;BRASS&#10;BRASS&#10;ORCHESTRA&#10;PIANO 1&#10;PIANO 4&#10;E. PIANO 1&#10;E. PIANO 1&#10;E. PIANO 2&#10;E. PIANO 3&#10;CLAVI&#10;HARPSICHORD&#10;HARPSICHORD&#10;VIBE&#10;DIGI BASS 1&#10;DIGI BASS 2&#10;DIGI BASS 2&#10;PICK BASS&#10;GLOCKEN&#10;GLOCKEN&#10;TINE&#10;TINE&#10;TINE&#10;TUBE BELL&#10;TUBE BELL&#10;TUBE BELL&#10;XYLOPHONE&#10;XYLOPHONE&#10;HARP&#10;KOTO&#10;SITAR&#10;SITAR&#10;KALIMBA&#10;KALIMBA&#10;KALIBMA&#10;LOG DRUM&#10;STEEL DRUM&#10;PIPE ORGAN 3&#10;PIPE ORGAN 3&#10;SYNTH 1&#10;SYNTH 2&#10;SYNTH 3&#10;SYNTH 3&#10;SYNTH 4&#10;SYNTH 4&#10;CLAVI&#10;DIGI BASS 1&#10;DIGI BASS 1&#10;PICK BASS&#10;PICK BASS&#10;PICK BASS&#10;ROUND BASS&#10;ROUND BASS&#10;HARMONICA&#10;HARP&#10;KOTO&#10;SITAR&#10;MARIMBA&#10;SYNTH 1&#10;BASS DRUM&#10;Ac SNARE&#10;TIGHT SNARE&#10;E. SNARE&#10;RIM&#10;Ac TOM&#10;H. HAT&#10;CRASH&#10;RIDE&#10;STRAT GUITAR&#10;FUZZ MUTE&#10;A. GUITAR&#10;F. GUITAR&#10;GUITAR HARMO&#10;PULL BASS&#10;BASS HARMO&#10;BOWED STRING&#10;STRING ATTACK&#10;STRING SUS&#10;PIZZICATO&#10;PIANO&#10;EL. GRAND&#10;PIANO NOISE&#10;TRUMPET&#10;SHAKUHACHI ATTACK&#10;SHAKUHACHI SUS&#10;PAN FLUTE ATTACK&#10;PAN FLUTE SUS&#10;VOICE&#10;WHITE NOISE &lt;LOOP&gt;&#10;STRING LOOP&#10;SHAKUHACHI LOOP&#10;PAN FLUTE LOOP&#10;VOICE LOOP&#10;WHITE NOISE LOOP&#10;Ac SNARE LOOP&#10;F. GUITAR LOOP&#10;PULL BASS LOOP&#10;OMNIBUS LOOP 1&#10;OMNIBUS LOOP 2&#10;OMNIBUS LOOP 3&#10;OMNIBUS LOOP 4&#10;OMNIBUS LOOP 5&#10;OMNIBUS LOOP 6&#10;OMNIBUS LOOP 7&#10;OMNIBUS LOOP 8 &lt;REVERSE&gt;&#10;Ac SNARE REV&#10;Ac TOM REV&#10;F. GUITAR REV&#10;H. HAT ALT&#10;CRASH ALT&#10;PIANO NOISE ALT"
             uiComboMenuBackgroundRibbed="1" uiComboButtonGradient="0" uiComboButtonGradientColour1="ff0000ff"
             uiComboButtonGradientColour2="ff00008b" uiComboButtonWidthOverride="0"
             uiComboButtonWidth="16" uiComboDynamicContent="0" uiComboSelectedId="-1"
             uiComboSelectedIndex="-1" componentBubbleRoundAngle="10.00000000000000000000"
             componentBubbleBackgroundColour1="0x9cffffff" componentBubbleBackgroundColour2="0xbab9b9b9"
             componentBubbleBackgroundGradientType="1" componentBubbleValueColour="0xff000000"
             componentBubbleValueFont="&lt;Sans-Serif&gt;;14;0;0;0;0;1;3"
             componentBubbleValueJustification="centred" componentBubbleNameColour="0xff000000"
             componentBubbleNameFont="&lt;Sans-Serif&gt;;14;0;0;0;0;1;3" componentBubbleNameJustification="centred"
             componentLayerUid="a90f05c882010000feffffffffffffff" componentRectangle="0 62 160 34"
             uiType="uiCombo"/>
```

So see we have uiComboSelectedId="-1"
             uiComboSelectedIndex="-1" 
You can comp::setProperty("uiComboSelectedIndex", "10")
and this will be like someone clicked comboBox entry 10 .. not sure if it starts at 0.

Hope it makes sense what I am driving at.

**Comment by callimero** [comment]
Thanks for that long answer. I guess I want somthing much simpler, but thats only a minor thing:

I just want one button send "1" as value and the other "-1" as value and the lua then makes:

`
patchnumber = patchnumber + value
`
givin me countn the patchnumber up or down on every click. (and roll over at max and 0)

After "stealing" the isPanelReady function (method?) from the Pro2 panel I also ironed out some init-bugs and all works ok for me.

**Reply by shooking** [reply]
I suspect someone will make a short alternative. Forget the length of the XML - that was just to show how you can use Lua to introspect status and dynamically change ANYTHING on the fly - just be careful of causing feedback loops.

What you describe I sort of do on my K1r panel. The inc/dec sends the sysex to the synth to load next patch from memory. I thought you wanted to choose the next item on a combo box that was being populated from reading a directory.

**Comment by dobo365** [comment]
Nice to read that you are using my Pro-2 panel for reference ;-)
Pay attention that not everything is perfect in it and that I will take the time during this year to bring it to the level of the current panels I'm making (much less methods, more efficient ones).

About the isPanelReady() method, check that the if condition is like
`	if panel:getBootstrapState() == false and bPanelLoaded == true then`
otherwise the panel will not work as VST/AU.

About your button request:
- I'm only using uiImageButton components and not the simple uiButton. They have different properties but the most interesting one is indeed that you can give it the look you want by just building a vertical image with 2 frames (I explain that in my Step by step guide)
- the uiImageButton has 2 default values, False and True but also a property Button mode that you should set on Momentary
- create 2 buttons and oly one OnChange method for your buttons. In your OnChange method you can track which button has been pressed and increase/decrease your patch number accordingly. In fact you don't care about the value of the button...

```
	-- No action if the panel is in bootstrap state
	if not isPanelReady() then
		return
	end	

	sName= L(mod:getName())

	-- Repaint corresponding Envelope graph
	if sName == "IncreaseButton" then 
		-- some actions
	elseif sName == "DecreaseButton" then 
		-- some actions
	end

end
```

You can also use uiSlider with type IncDecButton but you need to give some interval which I don't think is good for your case.
As indicated by @shooking you can modify component properties dynamically as well.
No need to look at the xml code ;-)
Just turn the Property IDs/Names on or off in the Ctrlr View menu. You need to click on another component  to see the effect.
You will then see the xml name of the properties rather than the explanatory text.
Then as indicated you can use comp::setProperty("uiLabelText", "Hello world")

Hope this helps

**Reply by shooking** [reply]
I didnt know that :-) Hence why I parse the XML direct and make a cheat sheet of common values for quick reference. Nice tip.

**Reply by callimero** [reply]
Using the name is clever :) So I was not that blindfolded and just did not find another way.

Inbetween I ironed some bugs out and experimented with some button graphics.


[RefaceCS_1_1_20200123_2021-01-15_19-28.zip](https://github.com/RomanKubiak/ctrlr/files/5822339/RefaceCS_1_1_20200123_2021-01-15_19-28.zip)
